### PR TITLE
Object Detection

### PR DESCRIPTION
**Pull request by warthecatalyst**
This PR is mainly about yolov3 model in ObjectDetection.
1. add Yolov3 Model in basic model zoo and can be trained with a simple loss function for Pytorch engine. And corresponding test.
2. add Yolov3 Loss function, maybe still have some bugs to fix
3. add an example of yolov3Training: TrainPikachuWithYOLOV3
4. About ObjectDetectionDataSet, add a new function getClasses() because for most ObjectDetection Models and Translators, we need to know how many types of objects in the dataset.
5. Add ObjectDetection for DJL-Zero, and due to small bugs for yolov3, currently just train a SingleShotDetection Model.

TODO List:
1. Translator of Yolov3(actually I'm written a demo but it can't pass the code check so I stash it.
2. For COCODataSet, we should update synset.

### CODE REVIEW

File: model-zoo/src/main/java/ai/djl/basicmodelzoo/cv/object_detection/yolo/YOLOV3.java [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+        //Outputs
+        return new NDList(
+                out0.singletonOrThrow(),
+                out1.singletonOrThrow(),
+                out2.singletonOrThrow()
+        );
+    }
+
+    @Override
+    public Shape[] getOutputShapes(Shape[] inputShapes) {
```
Comment (zachgk) [inline model-zoo/src/main/java/ai/djl/basicmodelzoo/cv/object_detection/yolo/YOLOV3.java:181]: Is this right? I would imagine that the logic here would be similar to forward if you are doing the computation fully, or could skip steps to replace computation with fixed knowledge. This looks a lot like a SequentialBlock logic, but I don't remember Yolo being sequential. Similar applies to the initializeChildBlocks which should also be similar to forward as it has to pass the right type of inputs to each intermediate block

File: api/src/main/java/ai/djl/training/loss/SingleShotDetectionLoss.java [hunk 2]
```diff
@@ -48,6 +48,8 @@ public SingleShotDetectionLoss() {
     @Override
     protected Pair<NDList, NDList> inputForComponent(
             int componentIndex, NDList labels, NDList predictions) {
+        System.out.println(labels.singletonOrThrow());  //print labels for test
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/training/loss/SingleShotDetectionLoss.java:51]: Remove testing code

File: api/src/main/java/ai/djl/training/loss/YOLOv3Loss.java [hunk 3]
```diff
@@ -0,0 +1,275 @@
+package ai.djl.training.loss;
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/training/loss/YOLOv3Loss.java:1]: Add license header

File: api/src/main/java/ai/djl/training/loss/YOLOv3Loss.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+import ai.djl.ndarray.types.DataType;
+import ai.djl.ndarray.types.Shape;
+import ai.djl.nn.Activation;
+import ai.djl.util.Pair;
+import ai.djl.util.PairList;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+
+public class YOLOv3Loss extends Loss{
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/training/loss/YOLOv3Loss.java:19]:
```
./gradlew formatJava
```

File: api/src/main/java/ai/djl/training/loss/YOLOv3Loss.java [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+
+        //to calculate the area of prediction bbox and true bbox
+        NDArray areaA = boxATrue.get(":,2").sub(boxATrue.get(":,0")).mul(boxATrue.get(":,3").sub(boxATrue.get(":,1"))).expandDims( 1).broadcast(inter.getShape()),
+                areaB = boxBTrue.get(":,2").sub(boxBTrue.get(":,0")).mul(boxBTrue.get(":,3").sub(boxBTrue.get(":,1"))).expandDims( 0).broadcast(inter.getShape());
+
+        NDArray union = areaA.add(areaB).sub(inter);
+        return inter.div(union);
+    }
+
+    public ArrayList<PairList<Long,Rectangle>> getTargetFromCurrentLabel(NDArray labels){
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/training/loss/YOLOv3Loss.java:257]: Use List instead of `ArrayList` in function declaration

File: api/src/main/java/ai/djl/training/loss/YOLOv3Loss.java [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+            return new NDList(iou,boxLossScale,groundTruth);
+        }
+
+
+
+        return null;
+    }
+
+    //calculate IOU is already defined in Rectangle
+    public NDArray calculateIOU(NDArray boxA,NDArray boxB){
```
Comment (frankfliu) [inline api/src/main/java/ai/djl/training/loss/YOLOv3Loss.java:211]: javadoc

File: examples/src/main/java/ai/djl/examples/training/TrainPikachu.java [hunk 7]
```diff
@@ -87,7 +87,7 @@ public static TrainingResult runExample(String[] args) throws IOException, Trans
             try (Trainer trainer = model.newTrainer(config)) {
                 trainer.setMetrics(new Metrics());
 
-                Shape inputShape = new Shape(arguments.getBatchSize(), 3, 256, 256);
+                Shape inputShape = new Shape(1, 3, 256, 256);
```
Comment (frankfliu) [inline examples/src/main/java/ai/djl/examples/training/TrainPikachu.java:90]: Why?

**Comment by warthecatalyst** [comment]
emm,this is just a test pr for my mentor Zack to get a view on my current progress and make suggestions



At 2022-08-29 13:50:35, "Frank Liu" ***@***.***> wrote:

@frankfliu commented on this pull request.

In api/src/main/java/ai/djl/training/loss/SingleShotDetectionLoss.java:

> @@ -48,6 +48,8 @@ public SingleShotDetectionLoss() {
     @Override
     protected Pair<NDList, NDList> inputForComponent(
             int componentIndex, NDList labels, NDList predictions) {
+        System.out.println(labels.singletonOrThrow());  //print labels for test


Remove testing code

In api/src/main/java/ai/djl/training/loss/YOLOv3Loss.java:

> @@ -0,0 +1,275 @@
+package ai.djl.training.loss;


Add license header

In api/src/main/java/ai/djl/training/loss/YOLOv3Loss.java:

> +import ai.djl.modality.cv.output.Rectangle;
+import ai.djl.ndarray.NDArray;
+import ai.djl.ndarray.NDArrays;
+import ai.djl.ndarray.NDList;
+import ai.djl.ndarray.NDManager;
+import ai.djl.ndarray.index.NDIndex;
+import ai.djl.ndarray.types.DataType;
+import ai.djl.ndarray.types.Shape;
+import ai.djl.nn.Activation;
+import ai.djl.util.Pair;
+import ai.djl.util.PairList;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+
+public class YOLOv3Loss extends Loss{

./gradlew formatJava


In api/src/main/java/ai/djl/training/loss/YOLOv3Loss.java:

> +                boxBTrue.get(":,2:").expandDims(0).broadcast(A,B,2));
+        NDArray minXY = NDArrays.maximum(boxATrue.get(":,:2").expandDims(1).broadcast(A,B,2),
+                boxBTrue.get(":,:2").expandDims(0).broadcast(A,B,2));
+
+        NDArray inter = NDArrays.minimum(maxXY.sub(minXY),0);
+        inter = inter.get(":,:,0").mul(inter.get(":,:,1"));
+
+        //to calculate the area of prediction bbox and true bbox
+        NDArray areaA = boxATrue.get(":,2").sub(boxATrue.get(":,0")).mul(boxATrue.get(":,3").sub(boxATrue.get(":,1"))).expandDims( 1).broadcast(inter.getShape()),
+                areaB = boxBTrue.get(":,2").sub(boxBTrue.get(":,0")).mul(boxBTrue.get(":,3").sub(boxBTrue.get(":,1"))).expandDims( 0).broadcast(inter.getShape());
+
+        NDArray union = areaA.add(areaB).sub(inter);
+        return inter.div(union);
+    }
+
+    public ArrayList<PairList<Long,Rectangle>> getTargetFromCurrentLabel(NDArray labels){


Use List instead of ArrayList in function declaration

In api/src/main/java/ai/djl/training/loss/YOLOv3Loss.java:

> +                    float interHeight = Math.min(trueBottom,predBottom)-Math.max(trueTop,predTop);
+                    float inter = interWidth*interHeight, union = wgt.get(i).getFloat()*inW*hgt.get(i).getFloat()*inH
+                            + anchors.get(j,0).getFloat()*anchors.get(j,1).getFloat()-inter;
+                    iou.set(new NDIndex(curIndex),inter/union);
+                }
+            }
+            return new NDList(iou,boxLossScale,groundTruth);
+        }
+
+
+
+        return null;
+    }
+
+    //calculate IOU is already defined in Rectangle
+    public NDArray calculateIOU(NDArray boxA,NDArray boxB){


javadoc

In examples/src/main/java/ai/djl/examples/training/TrainPikachu.java:

> @@ -87,7 +87,7 @@ public static TrainingResult runExample(String[] args) throws IOException, Trans
             try (Trainer trainer = model.newTrainer(config)) {
                 trainer.setMetrics(new Metrics());
 
-                Shape inputShape = new Shape(arguments.getBatchSize(), 3, 256, 256);
+                Shape inputShape = new Shape(1, 3, 256, 256);


Why?

File: basicdataset/src/main/java/ai/djl/basicdataset/cv/CocoDetection.java [hunk 8]
```diff
@@ -79,6 +80,12 @@ public PairList<Long, Rectangle> getObjects(long index) {
         return labels.get(Math.toIntExact(index));
     }
 
+    /** {@inheritDoc} */
+    @Override
+    public List<String> getClasses() {
+        return synset;
```
Comment (zachgk) [inline basicdataset/src/main/java/ai/djl/basicdataset/cv/CocoDetection.java:86]: It is better to fail clearly rather than have something partial that may confuse people that it works. Here, just throw an exception, specifically at UnsupportedOperationException to indicate that it is not yet implemented.

File: examples/src/main/java/ai/djl/examples/training/TrainPikachu.java (hunk 7 above)
Comment (zachgk) [reply examples/src/main/java/ai/djl/examples/training/TrainPikachu.java:90]: You probably shouldn't modify the example to use a fixed batch size of 1 rather than the variable version

File: integration/src/main/java/ai/djl/integration/tests/model_zoo/object_detection/YOLOv3Test.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+import ai.djl.training.initializer.Initializer;
+import ai.djl.training.loss.Loss;
+import ai.djl.translate.Batchifier;
+import ai.djl.util.PairList;
+
+import org.testng.annotations.Test;
+
+public class YOLOv3Test {
+    @Test
+    public void testDarkNet() {
```
Comment (zachgk) [inline integration/src/main/java/ai/djl/integration/tests/model_zoo/object_detection/YOLOv3Test.java:39]: You should add a test for the full yolov3 model as well

**Review by zachgk — CHANGES_REQUESTED** [review]

**Review by zachgk — APPROVED** [review]
